File: paper/trie.cc
#include "trie.h"
#include <iostream>
#include <stdlib.h>
#include <queue>
#include <utility>
#include <map>

static inline int idx(char x) { return x - 'a'; }

namespace paper {

// Initially, this node is empty
Trie::Trie() {
  for (int i=0; i<kNumLetters; i++)
    children_[i] = NULL;
  is_word_ = false;
  mark_ = 0;
}

Trie* Trie::AddWord(const char* wd) {
  if (!wd) return NULL;
  if (!*wd) {
    SetIsWord();
    return this;
  }
  int c = idx(*wd);
  if (!StartsWord(c))
    children_[c] = new Trie;
  return Descend(c)->AddWord(wd+1);
}

Trie::~Trie() {
  for (int i=0; i<kNumLetters; i++) {
    if (children_[i]) delete children_[i];
  }
}

size_t Trie::Size() {
  size_t size = 0;
  if (IsWord()) size++;
  for (int i=0; i<kNumLetters; i++) {
    if (StartsWord(i)) size += Descend(i)->Size();
  }
  return size;
}

size_t Trie::NumNodes() {
  int count = 1;
  for (int i = 0; i < kNumLetters; i++)
    if (StartsWord(i))
      count += Descend(i)->NumNodes();
  return count;
}

// static
bool Trie::ReverseLookup(const Trie* base, const Trie* child,
                         std::string* out) {
  if (base==child) return true;
  for (int i=0; i<kNumLetters; i++) {
    if (base->StartsWord(i) && ReverseLookup(base->Descend(i), child, out)) {
      *out = std::string(1,'a'+i) + *out;
      return true;
    }
  }
  return false;
}

// static
std::string Trie::ReverseLookup(const Trie* base, const Trie* child) {
  std::string out;
  ReverseLookup(base, child, &out);
  return out;
}

void Trie::SetAllMarks(unsigned mark) {
  if (IsWord()) Mark(mark);
  for (int i=0; i<kNumLetters; i++) {
    if (StartsWord(i)) Descend(i)->SetAllMarks(mark);
  }
}

Trie* Trie::FindWord(const char* wd) {
  if (!wd) return NULL;
  if (!*wd) {
    return IsWord() ? this : NULL;
  }
  int c = idx(*wd);
  if (!StartsWord(c)) return NULL;
  return Descend(c)->FindWord(wd + 1);
}

Trie* Trie::CreateFromFile(const char* filename) {
  Trie* t = new Trie;
  char line[80];
  FILE* f = fopen(filename, "r");
  if (!f) {
    fprintf(stderr, "Couldn't open %s\n", filename);
    delete t;
    return NULL;
  }

  while (!feof(f) && fscanf(f, "%s", line)) {
    t->AddWord(line);
  }
  fclose(f);

  return t;
}

}  // namespace paper

File: go/boggle/trie.go
// Copyright 2011 Google Inc. All Rights Reserved.
// Author: danvk@google.com (Dan Vanderkam)

package trie

import (
  "fmt"
  "bufio"
  "os"
)

type Trie struct {
  is_word_ bool
  mark_ uint
  children_ []*Trie
}

func (t *Trie) IsWord() bool {
  return t.is_word_
}

func (t *Trie) SetIsWord() {
  t.is_word_ = true
}

func (t *Trie) StartsWord(c byte) bool {
  return t.children_[c] != nil
}

func (t *Trie) Descend(c byte) *Trie {
  return t.children_[c]
}
func (t *Trie) DescendMarkable(c byte) *Trie {
  return t.children_[c]
}

func (t *Trie) GetMark() uint {
  return t.mark_
}

func (t *Trie) Mark(mark uint) {
  t.mark_ = mark;
}

func (t *Trie) DescendOrAdd(c byte) *Trie {
  if t.children_[c] == nil {
    t.children_[c] = newTrie()
  }
  return t.children_[c]
}


func newTrie() *Trie {
  t := new(Trie)
  t.children_ = make([]*Trie, 26)
  return t
}

func NewTrie() *Trie {
  return newTrie()
}

func AddWord(t *Trie, word string) *Trie {
  if word == "" {
    t.SetIsWord()
    return t
  }

  if word[0] < 'a' || word[0] > 'z' {
    return nil
  }

  child := t.DescendOrAdd(word[0] - 'a')
  return AddWord(child, word[1:])
}

func IsWord(t *Trie, word string) bool {
  child := DescendPath(t, word)
  if child == nil { return false }
  return child.IsWord()
}

func DescendPath(t *Trie, word string) *Trie {
  if word == "" {
    return t
  }

  if word[0] < 'a' || word[0] > 'z' {
    return nil
  }

  if !t.StartsWord(word[0] - 'a') {
    return nil
  }

  return DescendPath(t.Descend(word[0] - 'a'), word[1:])
}


func CountWords(t *Trie) int {
  count := 0
  if t.IsWord() {
    count += 1
  }
  var i uint8
  for i = 0; i < 26; i++ {
    if t.StartsWord(i) {
      count += CountWords(t.Descend(i))
    }
  }
  return count
}

func PrintTrie(t *Trie, prefix string) {
  if t.IsWord() {
    fmt.Printf("%s\n", prefix)
  }

  var i uint8;
  for i = 0; i < 26; i++ {
    if t.StartsWord(i) {
      PrintTrie(t.Descend(i), prefix + string('a' + i))
    }
  }
}


func ReverseLookup(base, node *Trie) string {
  found, str := reverseLookup(base, node, "")
  if found {
    return str
  }
  return "?"
}

func reverseLookup(base, node *Trie, prefix string) (bool, string) {
  if base == node { return true, prefix }

  var i uint8
  for i = 0; i < 26; i++ {
    if base.StartsWord(i) {
      found, str := reverseLookup(base.Descend(i), node, prefix + string('a' + i))
      if found {
        return found, str
      }
    }
  }
  return false, ""
}


func CreateTrieFromFile(r *bufio.Reader) *Trie {
  t := NewTrie()
  for {
    line, isPrefix, err := r.ReadLine()
    if err == os.EOF {
      break
    }
    if isPrefix {
      fmt.Println("need more buffering!")
      return nil
    }
    if err != nil {
      fmt.Println("error! ", err)
      return nil
    }

    word := string(line)
    AddWord(t, word)
  }

  return t
}

func CreateTrieFromFilename(filename string) *Trie {
  f, err := os.Open(filename)
  if err != nil {
      fmt.Println(err)
      return nil
  }
  defer f.Close()

  r, err := bufio.NewReaderSize(f, 4*1024)
  if err != nil {
      fmt.Println(err)
      return nil
  }
  return CreateTrieFromFile(r)
}
